src/NHibernate/IStatelessSession.cs

```diff
@@ -143,6 +143,15 @@ public partial interface IStatelessSession : IDisposable
 		/// <summary>Close the stateless session and release the ADO.NET connection.</summary>
 		void Close();
 
+		/// <summary>
+		/// Cancel execution of the current query.
+		/// </summary>
+		/// <remarks>
+		/// May be called from one thread to stop execution of a query in another thread.
+		/// Use with care!
+		/// </remarks>
+		void CancelQuery();
```

Comment:
Adding a new member to an interface is a binary breaking change. These kind of changes have to be postponed until the next major version.

As a workaround, we use extension methods to avoid the breaking change, with a comment telling to do the change on the next major. See the extension class at the start of this file for examples.

Reply:
Okay, I moved it to an extension class.

Reply:
ISession don't have test  for ISession.CancelQuery(). Why IStatelessSession.CancelQuery() must have it?

Test hould be like this(in PostgreSQL), I suppose:
```C#

[Test]
public async Task CancelQuery()
{
    IStatelessSession ss = Sfi.OpenStatelessSession();
    var queryTask = Task.Factory.StartNew(() =>
    {
        var result = ss.CreateSQLQuery(@"
SELECT 
  clock_timestamp(),
  pg_sleep(10);
").UniqueResult();
    });

    Thread.Sleep(2000);
    ss.CancelQuery();
    try
    {
        await queryTask;
    }
    catch (Exception ex)
    {
        Assert.IsNotNull(ex.InnerException);
	Assert.AreEqual(ex.InnerException.GetType(), typeof(OperationCanceledException));
	var canceledException = (OperationCanceledException) ex.InnerException;
        Assert.IsNotNull(canceledException.InnerException);
	var postgresException = canceledException.InnerException as PostgresException;
	Assert.IsNotNull(postgresException);
	Assert.AreEqual(postgresException.SqlState, PostgresErrorCodes.QueryCanceled);
    }
}
```
We need to be sure of the request execution time in order to have time to cancel it(' pg_sleep(10);').
The main problem is that this query is related to PostgreSQL, just like the error code.

Do we need a test that checks for the DbCommand.Cancel() call? This is not the responsibility of Hibernate?
In AbstractBatcher.cs:
```C#
public void CancelLastQuery()
		{
			try
			{
				if (_lastQuery != null)
				{
					_lastQuery.Cancel();
				}
			}
			catch (HibernateException)
			{
				// Do not call Convert on HibernateExceptions
				throw;
			}
			catch (Exception sqle)
			{
				throw Convert(sqle, "Could not cancel query");
			}
		}
```

Reply:
Development practices change over time. Not everything has been developed with proper tests in the past, but that is not a valid reason for adding features or fixing bugs without corresponding tests.

About your proposal, we may accept a test specific to some databases, when devising one working for all is not practical, provided the test runs for at least one of the databases we have automated tests for. So a test specific to PostgreSql can be accepted. Use Assert.Ignore to skip other databases.

But maybe you can test it in a more database agnostic way, by using another session to exclusively lock some data, and then perform queries on the same data with exclusive locks too with another session on which `CancelQuery` will be called.

The main pitfall with your change is, does the session always use the batcher to run queries? If not, calling that function would do nothing, and that would be a bug. Ideally all ways of querying with a session should have a corresponding `CancelQuery` test.

Reply:
If it's a select query (not an insert/update/delete command), then yes, it always uses Batcher when calling "List" or "List<T>" (as far as I can see).
Added tests.